shopfloor_delivery_shipment/services/delivery_shipment.py

```diff
@@ -593,46 +582,87 @@ def _data_for_content_to_load_from_pickings(self, shipment_advice):
     ):
         """Return a dictionary where keys are source locations
         and values are dictionaries listing package_levels and move_lines
         loaded or to load.
+        """
+        # Grab move lines to sort, restricted to the current delivery
+        if picking:
+            move_lines = self._find_move_lines_to_process_from_picking(
+                shipment_advice, picking
+            )
+        elif location:
+            move_lines = self._find_move_lines_from_location(shipment_advice, location)
+        group_by_sale = bool(location)
+        return self._prepare_data_for_content(move_lines, group_by_sale=group_by_sale)
+
+    def _prepare_data_for_content(self, move_lines, group_by_sale=False):
+        """Returns data for package levels and/or move lines.
+
+        The data is grouped by source location with a key for package levels and
+        a key for move lines.
+        The package levels can be grouped by related sales order and if they are not
+        they will be all stored into a key whose name is only spaces.
+        The move lines are not grouped by sales.
 
         E.g:
             {
                 "SRC_LOCATION1": {
-                    "package_levels": [{PKG_LEVEL_DATA}, ...],
+                    "package_levels": {
+                       "sale 01": [{PKG_LEVEL_DATA}, ...],
+                       "sale 02": [{PKG_LEVEL_DATA}, ...],
+                       " ": [{PKG_LEVEL_DATA}, ...],
+                    }
                     "move_lines": [{MOVE_LINE_DATA}, ...],
                 },
                 "SRC_LOCATION2": {
                     ...
                 },
             }
         """
-        # Grab move lines to sort, restricted to the current delivery
-        if picking:
-            move_lines = self._find_move_lines_to_process_from_picking(
-                shipment_advice, picking
-            )
-        elif location:
-            move_lines = self._find_move_lines_from_location(shipment_advice, location)
-        return self._prepare_data_for_content(move_lines)
-
-    def _prepare_data_for_content(self, move_lines):
+        empty_group_name = "null"
         data = collections.OrderedDict()
         package_level_ids = []
+
         # Sort and group move lines by source location and prepare the data
-        for move_line in move_lines.sorted(lambda ml: ml.location_id.name):
+        sales_started = []
+        if group_by_sale:
+            sales_started = move_lines.shipment_advice_id.loaded_picking_ids.sale_id
+            # Grouping by sales (the packages) sort starting from last position
+            #  2 the lines NOT in a sales order whose loading has started.
+            #  1 the lines in a sales being loaded and the line is done.
+            #  0 the lines in a sales being loaded and not yet done
+            move_lines = move_lines.sorted(
+                lambda ml: 2
+                if ml.move_id.sale_line_id.order_id not in sales_started
```

Comment:
We need to add `sale_stock` as dependency, Shopfloor doesn't depend on it, better to limit this change to `shopfloor_delivery_shipment`.

Reply:
Seems that `shopfloor_reception` has also tests depending on `sale` (while the module doesn't really depend on this module to work), this would need some tests rewrite. Hopefully some modules in OCA/wms are installing `sale` so tests are running well, but isolated they should crash.

Reply:
The module `shipment_advice` depends on `delivery` which depends on `sale_stock`